lib/lookup/src/lookup_v2/compat.rs

```diff
@@ -0,0 +1,79 @@
+///! Contains backwards compatibility with lookup "v1"
+use crate::lookup_v2::{BorrowedSegment, Path};
+use crate::{LookupBuf, SegmentBuf};
+use std::borrow::Cow;
+
+impl<'a> Path<'a> for &'a LookupBuf {
+    type Iter = LookupBufPathIter<'a>;
+
+    fn segment_iter(&self) -> Self::Iter {
+        LookupBufPathIter {
+            buf: self,
+            segment_i: 0,
+            coalesce_i: 0,
+        }
+    }
+}
+
+pub struct LookupBufPathIter<'a> {
+    buf: &'a LookupBuf,
+    segment_i: usize,
+    coalesce_i: usize,
+}
+
+impl<'a> Iterator for LookupBufPathIter<'a> {
+    type Item = BorrowedSegment<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.buf.segments.get(self.segment_i) {
+            Some(SegmentBuf::Field(field)) => {
+                self.segment_i += 1;
+                Some(BorrowedSegment::Field(Cow::Borrowed(&field.name)))
+            }
+            Some(SegmentBuf::Index(index)) => {
+                self.segment_i += 1;
+                Some(BorrowedSegment::Index(*index))
+            }
+            Some(SegmentBuf::Coalesce(fields)) => match fields.get(self.coalesce_i) {
+                Some(field) => {
+                    if self.coalesce_i == fields.len() - 1 {
+                        self.coalesce_i = 0;
+                        self.segment_i += 1;
+                        Some(BorrowedSegment::CoalesceEnd(Cow::Borrowed(&field.name)))
+                    } else {
+                        self.coalesce_i += 1;
+                        Some(BorrowedSegment::CoalesceField(Cow::Borrowed(&field.name)))
+                    }
+                }
+                None => unreachable!(),
+            },
+            None => None,
+        }
```

Comment:
This looks like it would be a little simpler with a `map`:
```suggestion
        self.buf.segments.get(self.segment_i).map(|segment| match segment {
            SegmentBuf::Field(field) => {
                self.segment_i += 1;
                BorrowedSegment::Field(Cow::Borrowed(&field.name))
            }
            SegmentBuf::Index(index) => {
                self.segment_i += 1;
                BorrowedSegment::Index(*index)
            }
            SegmentBuf::Coalesce(fields) => match fields.get(self.coalesce_i) {
                Some(field) => {
                    if self.coalesce_i == fields.len() - 1 {
                        self.coalesce_i = 0;
                        self.segment_i += 1;
                        BorrowedSegment::CoalesceEnd(Cow::Borrowed(&field.name))
                    } else {
                        self.coalesce_i += 1;
                        BorrowedSegment::CoalesceField(Cow::Borrowed(&field.name))
                    }
                }
                None => unreachable!(),
            })
```

Reply:
Also, let's add a comment to `unreachable` to explain why it is unreachable:

```rust
unreachable!("path parser ensures coalesced segments always include at least one field")
```